internal/native_image/rules.bzl

```diff
@@ -84,15 +96,14 @@ def _graal_binary_implementation(ctx):
         direct_inputs,
         transitive = transitive_inputs,
     )
-
     run_params = {
         "outputs": [binary],
         "executable": graal,
         "inputs": inputs,
         "mnemonic": "NativeImage",
         "env": native_toolchain.env,
         "execution_requirements": {k: "" for k in native_toolchain.execution_requirements},
-        "progress_message": "Compiling native image %{label}",
+        "progress_message": "Native Image (__mode__) %{label}".replace("__mode__", _build_action_message(ctx)),
```

Comment:
This will still end up creating a new retained string per target. Instead, you could have a dict of fixed strings per mode.